add-custom-dynamic-entity-data-source/src/main/java/com/esri/arcgismaps/sample/addcustomdynamicentitydatasource/components/CustomEntityFeedProvider.kt

```diff
@@ -0,0 +1,172 @@
+                withContext(NonCancellable) {
+                    _feed.tryEmit(
+                        CustomDynamicEntityDataSource.FeedEvent.ConnectionFailure(
+                            e,
+                            true
+                        )
+                    )
+                }
+            }
+        }
+    }
+
+    /**
+     * Processes the given line from the observations file and returns a new observation.
+     */
+    private fun processNextObservation(readLine: String): CustomDynamicEntityDataSource.FeedEvent.NewObservation {
+
+        // Get the next observation from the file and parse it as a JSON object.
+        val jsonElement = Json.parseToJsonElement(readLine)
+
+        // Get the x and y coordinates of the observation.
+        val point =
+            (jsonElement.jsonObject["geometry"] as? JsonObject)?.let { geometryJsonObject ->
+                // Create a new MapPoint from the x and y coordinates of the observation.
+                Point(
+                    geometryJsonObject["x"]!!.jsonPrimitive.double,
+                    geometryJsonObject["y"]!!.jsonPrimitive.double
+                )
+            }
+        // Get the dictionary of attributes from the observation using the field names as keys.
+        val attributes = mutableMapOf<String, Any?>()
+        (jsonElement.jsonObject["attributes"] as? JsonObject)?.let { attributesJsonObject ->
+            attributesJsonObject.entries.forEach { (key, value) ->
+                if (value is JsonPrimitive) {
+                    attributes[key] = value.contentOrNull
+                }
+            }
+        }
+        // Return a new observation with the point and attributes.
+        return CustomDynamicEntityDataSource.FeedEvent.NewObservation(
+            point,
+            attributes
+        )
+    }
+
+    /**
+     * Returns the schema for the custom data source.
+     */
+    private fun getSchema(): List<Field> {
+        // Return a list of fields matching the attributes of each observation in the custom data source.
+        return listOf(
+            Field(FieldType.Text, "MMSI", "", 256),
+            Field(FieldType.Float64, "BaseDateTime", "", 8),
+            Field(FieldType.Float64, "LAT", "", 8),
+            Field(FieldType.Float64, "LONG", "", 8),
+            Field(FieldType.Float64, "SOG", "", 8),
+            Field(FieldType.Float64, "COG", "", 8),
+            Field(FieldType.Float64, "Heading", "", 8),
+            Field(FieldType.Text, "VesselName", "", 256),
+            Field(FieldType.Text, "IMO", "", 256),
+            Field(FieldType.Text, "CallSign", "", 256),
+            Field(FieldType.Text, "VesselType", "", 256),
+            Field(FieldType.Text, "Status", "", 256),
+            Field(FieldType.Float64, "Length", "", 8),
+            Field(FieldType.Float64, "Width", "", 8),
+            Field(FieldType.Text, "Cargo", "", 256),
+            Field(FieldType.Text, "globalid", "", 256)
+        )
+    }
```

Comment:
I suggest to define this as a const or lazy property. I think a lazy property would be fine here:
```suggestion
   private val schema: List<Field> by lazy { 
        listOf(
            Field(FieldType.Text, "MMSI", "", 256),
            Field(FieldType.Float64, "BaseDateTime", "", 8),
            Field(FieldType.Float64, "LAT", "", 8),
            Field(FieldType.Float64, "LONG", "", 8),
            Field(FieldType.Float64, "SOG", "", 8),
            Field(FieldType.Float64, "COG", "", 8),
            Field(FieldType.Float64, "Heading", "", 8),
            Field(FieldType.Text, "VesselName", "", 256),
            Field(FieldType.Text, "IMO", "", 256),
            Field(FieldType.Text, "CallSign", "", 256),
            Field(FieldType.Text, "VesselType", "", 256),
            Field(FieldType.Text, "Status", "", 256),
            Field(FieldType.Float64, "Length", "", 8),
            Field(FieldType.Float64, "Width", "", 8),
            Field(FieldType.Text, "Cargo", "", 256),
            Field(FieldType.Text, "globalid", "", 256)
        )
    }
```